// t2p-clientlib/index.js
const dateFormat = require('dateformat');
var crypto = require('crypto');

var testUriH = '/authen/v1/clientToken/testClientToken';


function generateTokenTypeH(bodyString, clientKey, endpointURI = '/authen/v1/clientToken/generate', isEncrypted = true) 
{
    if(typeof bodyString === "undefined")
    {
        return false;
    }
    
    if(typeof clientKey === "undefined")
    {
        return false;
    }

    keys = parseKey(clientKey)
    headerInfo = prepareHeaderInfo(keys, "H", endpointURI, false)
    encryptedBody=bodyString
    if(bodyString!="" && isEncrypted) {
        encryptedBody = encryptBody(bodyString, keys)   
    } 
    headerString = prepareHeader(headerInfo, encryptedBody)
    hashmac = createHeaderHashMac(headerString, keys)
    // console.log(headerString)
    // console.log(hashmac)
    // console.log(headerInfo)
    
    result = {}
    result.header = Buffer.from(Buffer.from(JSON.stringify(headerInfo)).toString('base64')+':'+hashmac).toString('base64')
    result.body = encryptedBody
    return result;
}

function prepare(bodyString, clientKey, endpointURI = '/authen/v1/clientToken/generate', testMode = false) 
{
    if(typeof bodyString === "undefined")
    {
        return false;
    }
    
    if(typeof clientKey === "undefined")
    {
        return false;
    }

    keys = parseKey(clientKey)
    headerInfo = prepareHeaderInfo(keys, "H", endpointURI, testMode)
    encryptedBody=""
    if(bodyString!="") {
        encryptedBody = encryptBody(bodyString, keys)   
    } 
    headerString = prepareHeader(headerInfo, encryptedBody)
    hashmac = createHeaderHashMac(headerString, keys)
    // console.log(headerString)
    // console.log(hashmac)
    // console.log(headerInfo)
    
    result = {}
    result.header = Buffer.from(Buffer.from(JSON.stringify(headerInfo)).toString('base64')+':'+hashmac).toString('base64')
    result.body = encryptedBody
    return result;
}

function prepareHS(bodyString, clientKey, endpointURI = '/authen/v1/clientToken/generateNew')
{
    if(typeof bodyString === "undefined")
    {
        return false;
    }
    if(typeof clientKey === "undefined")
    {
        return false;
    }

    keys = parseKey(clientKey)
    headerInfo = prepareHeaderInfo(keys, "HS", endpointURI) 
    headerString = prepareHeader(headerInfo, bodyString)
    hashmac = createHeaderHashMac(headerString, keys)

    result = {}
    result.header = Buffer.from(Buffer.from(JSON.stringify(headerInfo)).toString('base64')+':'+hashmac).toString('base64')
    result.body = bodyString
    return result;
}

function createHeaderHashMac(headerStr, keys)
{
    // console.log(keys.hashKey)
    key = Buffer.from(keys.hashKey,'base64')
    var hash = crypto.createHmac('sha512', key)
    hash.update(headerStr)
    var res = hash.digest('hex')
    
    return res
}

function encryptBody(bodyString, keys)
{
    // var token = crypto.randomBytes(64).toString('hex');
    var aesKey = crypto.randomBytes(32);
    var ivKey = crypto.randomBytes(16);    
    
    var decipher = crypto.createCipheriv('aes-256-cbc', aesKey, ivKey) 
    var enc = Buffer.concat([
                        decipher.update(bodyString),
                        decipher.final()
                    ]);
    
    var encryptedKey = encryptKey(Buffer.from(aesKey).toString('base64')+':'+Buffer.from(ivKey).toString('base64'), keys.publicKey)                    
    var finalEncrypted = encryptedKey.toString("base64")+":"+enc.toString("base64")
    return finalEncrypted
}

function decryptBody(bodyString, keys)
{
    if(!bodyString.includes(":")){
        return false;
    }
    var bodyString = bodyString.split(':')        
    var decryptedKey = decryptKey(bodyString[0], keys.privateKey)

    var aesKey = decryptedKey.toString().split(':')       
    var decipher = crypto.createDecipheriv('aes-256-cbc', Buffer.from(aesKey[0],'base64'), Buffer.from(aesKey[1],'base64')) 
    var dec = Buffer.concat([
                        decipher.update(Buffer.from(bodyString[1],'base64')),
                        decipher.final()
                    ]);
    return dec.toString()
}


function encryptKey(aesKeyIv, publicKey)
{
    // console.log(publicKey)
    var encryptedData = crypto.publicEncrypt(
        {
            key: publicKey,
            padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
            oaepHash: "sha1",
        },
        // We convert the data string to a buffer using `Buffer.from`
        Buffer.from(aesKeyIv)
    )
    return encryptedData
}

function decryptKey(aesKeyIv, privateKey)
{
    // console.log(publicKey)
    var decryptedText = crypto.privateDecrypt(
        {
            key: privateKey,
            padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
            oaepHash: "sha1",
        },
        // We convert the data string to a buffer using `Buffer.from`
        Buffer.from(aesKeyIv,'base64')
    )
    return decryptedText
}

function prepareHeader(headerInfo, encryptedBody)
{
    var concat=''
    var orderedHeader = Object.keys(headerInfo).sort().reduce(
        (obj, key) => { 
          concat  = concat+headerInfo[key]; 
          return concat;
        }, 
        {}
      );
    return orderedHeader+encryptedBody
}

function prepareHeaderInfo(keys, tokeyType, endpointURI, isTestMode)
{
    headerInfo = {};
    headerInfo.clientCode = keys.clientCode;
    headerInfo.keyCode = keys.keyCode;
    headerInfo.timestamp = dateFormat(new Date(), "yyyymmddHHMMss"); 
    headerInfo.clientLibVersion = "1.0.0";
    headerInfo.tokenType = tokeyType;
    
    headerInfo.uri = endpointURI;
    if(isTestMode){
        headerInfo.uri = testUriH;
    }
    headerInfo.method = "POST";
    
    return headerInfo
}

function parseKey(clientKey)
{
    
    if(!clientKey.includes(":")){
        clientKey=Buffer.from(clientKey, 'base64').toString('utf8')
    }
    const keys = clientKey.split(':')
    result = {}

    result.clientCode = keys[0].trim()    
    result.keyCode = keys[1].trim()    
    result.hashKey = keys[2].trim()
    result.publicKey = keys[3].trim()
    result.privateKey = keys[4].trim()

    return result
    
}

function encryptData(plainText, keyContent) {
    key = parseKey(keyContent)
    return encryptBody(plainText, key)
}

function decryptData(encryptedText, keyContent) {
    key = parseKey(keyContent)
    return decryptBody(encryptedText, key)
}

function generateHashMac(inputText, keyContent){
    keys = parseKey(keyContent)
    return createHeaderHashMac(inputText,keys)
}

function verifyHashMac(inputText, hashMac, keyContent){
    key = parseKey(keyContent)
    return createHeaderHashMac(inputText,keys)==hashMac
}

exports.preapare = prepare;
exports.preapareHS = prepareHS;
exports.prepare = prepare;
exports.prepareHS = prepareHS;
exports.encryptData = encryptData;
exports.decryptData = decryptData;
exports.generateHashMac = generateHashMac;
exports.verifyHashMac = verifyHashMac;
exports.generateTokenTypeH=generateTokenTypeH
